content/posts/2021-05-airdrop-anywhere-part-1.md

```diff
@@ -0,0 +1,65 @@
+
+<img src="/img/airdrop-anywhere-1.png" width=480 alt="Installing npcap"><br/>
+<sub style="color:lightgray">Installing npcap</sub>
+
+I was about to sit down and write the [P/Invoke](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke) code necessary to get .NET talking to this native library. That usually involves repetitive dissection of the header files provided with the library to construct the appropriate type signatures but, thankfully, the hard work has already been done by the maintainers of [SharpPcap](https://github.com/chmorgan/sharppcap). This library provides us with a way of initiating a packet capture and manipulating packets as we see fit. Yay!
+
+Additionally, there's [Packet.NET](https://github.com/chmorgan/packetnet) which provides a way to parse the raw packets that are being captured. This should be very helpful in allowing us to put together the protocol implementation needed for AWDL.
+
+But... this is where I hit a stumbling block - both my Windows desktop and laptop machines have wireless adapters but neither support [monitor mode](https://en.wikipedia.org/wiki/Monitor_mode) (aka RFMON) and whilst I could go and buy one easily enough, it appears that supporting monitor mode is not commonplace on Windows. Apple have gone the extra mile in shipping wireless hardware allows the AirDrop experience to be as seamless as possible but, sadly, that makes writing something that could be used by anybody on a Windows platform practically impossible. Eurgh.
+
+### Taking a couple of steps back
+
+At this point I started to consider the project a dead-end and just as I was switching off the laptop I thought - surely AirDrop is supported on wired adapters too? Turns out [it is](https://www.lifewire.com/airdrop-with-without-wifi-connection-2259801) but with some caveats - this is a legacy feature and AirDrop on any Mac that executes the following command is only available on networks that the Mac is connected to - no adhoc peer-to-peer fun!
+
+```
+defaults write com.apple.NetworkBrowser BrowseAllInterfaces 1
+```
+
+But, for our case of operating on Windows, that might be the glimmer of hope I'm looking for! The [OWL project](https://owlink.org) actually has an implementation of the services needed by AirDrop using the [Bonjour protocol](https://en.wikipedia.org/wiki/Bonjour_(software)).  Apple [provides support](https://developer.apple.com/bonjour/) for running Bonjour on Windows but it'd be nice to have a service that could run on any platform supported by .NET. While there's a whole bunch of libraries out there that support _browsing_ Bonjour-based services there's only one that seems to be maintained for _advertising_ services using it. Under the hood Bonjour is really an implementation of DNS-SD which uses multicast DNS as its basis. [Richard Schneider](https://github.com/richardschneider/net-mdns) has implemented an mDNS library that supports advertising services - this might be the magic we're looking for!
```

Comment:
```suggestion
But, for our case of operating on Windows, that might be the glimmer of hope I'm looking for! The [OWL project](https://owlink.org) actually has an implementation of the services needed by AirDrop using the [Bonjour protocol](https://en.wikipedia.org/wiki/Bonjour_(software)).  Apple [provides support](https://developer.apple.com/bonjour/) for running Bonjour on Windows but it'd be nice to have a service that could run on any platform supported by .NET. While there are a whole bunch of libraries out there that support _browsing_ Bonjour-based services there's only one that seems to be maintained for _advertising_ services using it. Under the hood Bonjour is really an implementation of DNS-SD which uses multicast DNS as its basis. [Richard Schneider](https://github.com/richardschneider/net-mdns) has implemented an [mDNS](https://en.wikipedia.org/wiki/Multicast_DNS) library that supports advertising services - this might be the magic we're looking for!
```